### PR TITLE
consensus: proto msgs

### PR DESCRIPTION
**Pull request by marbar3778**
## Description

migrate consensus reactor to protobuf

in places we are amino marshling json, the replacement for this will come in a secondary PR as this on has hit 2k changes.

Closes: #XXX

---

For contributor use:

- [ ] Wrote tests
- [ ] Updated CHANGELOG_PENDING.md
- [ ] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [ ] Updated relevant documentation (`docs/`) and code comments
- [ ] Re-reviewed `Files changed` in the Github PR explorer
- [ ] Applied Appropriate Labels

### CODE REVIEW

File: consensus/codec.go [hunk 1]
```diff
@@ -1,8 +1,15 @@
 package consensus
 
 import (
+	"fmt"
+
+	"github.com/gogo/protobuf/proto"
+	"github.com/pkg/errors"
```
Comment (erikgrinaker) [inline consensus/codec.go:7]: We should use the stdlib error handling.

File: proto/consensus/msgs.proto [hunk 2]
```diff
@@ -8,15 +8,15 @@ import "proto/types/types.proto";
...
 
 message NewRoundStep {
-  int64 height                   = 1;
-  int32 round                    = 2;
-  int32 step                     = 3 [(gogoproto.casttype) = "int8"];
-  int64 seconds_since_start_time = 4;
-  int32 last_commit_round        = 5;
+  int64  height                   = 1;
+  int32  round                    = 2;
+  int32 step                     = 3 [(gogoproto.casttype) = "uint8"];
```
Comment (erikgrinaker) [inline proto/consensus/msgs.proto:13]: Misaligned

File: crypto/merkle/simple_proof.go [hunk 3]
```diff
@@ -117,6 +118,30 @@ func (sp *SimpleProof) StringIndented(indent string) string {
...
+
+	pb.Total = sp.Total
+	pb.Index = sp.Index
+	pb.LeafHash = sp.LeafHash
+	pb.Aunts = sp.Aunts
+
+	return pb
+}
+
+func (sp *SimpleProof) FromProto(pb *tmcrypto.SimpleProof) error {
```
[the remark this answers is not in the document]
Comment (marbar3778) [reply crypto/merkle/simple_proof.go:132]: this approach makes sense. I was doing the other way to keep it grouped but there isn't a need for non-empty structs

File: proto/consensus/msgs.proto (hunk 2 above)
Comment (marbar3778) [reply proto/consensus/msgs.proto:13]: whoops I thought I opened this as a draft, not finished yet

File: libs/math/safemath.go [hunk 4]
```diff
@@ -54,6 +54,17 @@ func SafeConvertUint32(a int64) uint32 {
 	return uint32(a)
 }
 
+// SafeConvertUint32 takes an int64 and checks if it overflows
+// If there is an overflow it returns an error
+func SafeConvertUint8(a int64) (uint8, error) {
+	if a > math.MaxUint32 {
```
Comment (melekes) [inline libs/math/safemath.go:60]:
```suggestion
	if a > math.MaxUint8 {
```

File: proto/consensus/msgs.proto [hunk 5]
```diff
@@ -8,15 +8,15 @@ import "proto/types/types.proto";
...
 
 message NewRoundStep {
-  int64 height                   = 1;
-  int32 round                    = 2;
-  int32 step                     = 3 [(gogoproto.casttype) = "int8"];
-  int64 seconds_since_start_time = 4;
-  int32 last_commit_round        = 5;
+  int64  height                   = 1;
+  int32  round                    = 2;
+  uint32 step                     = 3;
```
Comment (melekes) [inline proto/consensus/msgs.proto:13]: why not uint8?

File: scripts/wal2json/main.go [hunk 6]
```diff
@@ -47,7 +47,17 @@ func main() {
 			panic(fmt.Errorf("failed to decode msg: %v", err))
 		}
 
-		json, err := cdc.MarshalJSON(msg)
+		pbWal, err := cs.WALToProto(msg.Msg)
+		if err != nil {
+			panic(fmt.Errorf("failed to transform walMessage into proto"))
```
Comment (melekes) [inline scripts/wal2json/main.go:52]:
```suggestion
			panic("failed to transform walMessage into proto")
```

File: consensus/codec.go [hunk 7]
```diff
@@ -1,15 +1,383 @@
...
+	case msgInfo:
+		consMsg, err := MsgToProto(msg.Msg)
+		if err != nil {
+			return nil, err
+		}
+		pb = tmcons.WALMessage{
+			Sum: &tmcons.WALMessage_MsgInfo{
+				MsgInfo: &tmcons.MsgInfo{
+					Msg:    *consMsg,
+					PeerId: string(msg.PeerID),
```
Comment (melekes) [inline consensus/codec.go:312]:
```suggestion
					PeerID: string(msg.PeerID),
```

?

File: proto/consensus/msgs.proto (hunk 5 above)
Comment (marbar3778) [reply proto/consensus/msgs.proto:13]: someone could send a number greater that uint8 over the wire then on unmarshal it would overflow panic.

**Pull request merged**